File: src/runtime-phase.js
const freezeRecord = (record) => Object.freeze(record);

const freezeRecordList = (records) =>
  Object.freeze(records.map((record) => freezeRecord(record)));

const normalizeNonEmptyToken = (value, label) => {
  if (typeof value !== "string") {
    throw new TypeError(`${label} must be a string`);
  }

  const normalizedValue = value.trim().toLowerCase();

  if (!normalizedValue) {
    throw new TypeError(`${label} must not be empty`);
  }

  return normalizedValue;
};

const normalizeAgentId = (agentId) => {
  if (typeof agentId !== "string") {
    throw new TypeError("agentId must be a string");
  }

  const normalizedAgentId = agentId.trim();

  if (!normalizedAgentId) {
    throw new TypeError("agentId must not be empty");
  }

  return normalizedAgentId;
};

const normalizeNonNegativeNumber = (value, label) => {
  if (typeof value !== "number" || Number.isNaN(value) || value < 0) {
    throw new TypeError(`${label} must be a non-negative number`);
  }

  return value;
};

const normalizePositiveInteger = (value, label) => {
  if (typeof value !== "number" || !Number.isInteger(value) || value <= 0) {
    throw new TypeError(`${label} must be a positive integer`);
  }

  return value;
};

const normalizeNonNegativeInteger = (value, label) => {
  if (typeof value !== "number" || !Number.isInteger(value) || value < 0) {
    throw new TypeError(`${label} must be a non-negative integer`);
  }

  return value;
};

export const RUNTIME_AUTHORIZED_IDLE_PHASES = Object.freeze([
  "idle",
  "rest",
  "break",
  "sleep",
]);

const EXPLICIT_PHASE_AUTHORITIES = new Set(["caller"]);
const RUNTIME_AUTHORIZED_IDLE_PHASE_SET = new Set(
  RUNTIME_AUTHORIZED_IDLE_PHASES,
);

export const createRuntimePhase = (value, options = {}) =>
  freezeRecord({
    value: normalizeNonEmptyToken(value, "runtime phase"),
    authority: normalizeNonEmptyToken(
      options.authority ?? options.source ?? "caller",
      "runtime phase authority",
    ),
    changedAt: options.changedAt ?? null,
    note: options.note ?? null,
  });

const normalizeRuntimePhase = (runtimePhase) => {
  if (typeof runtimePhase === "string") {
    return createRuntimePhase(runtimePhase);
  }

  if (!runtimePhase || typeof runtimePhase !== "object") {
    return null;
  }

  return createRuntimePhase(
    runtimePhase.value ?? runtimePhase.phase ?? runtimePhase.name,
    runtimePhase,
  );
};

export const createIdleWindowSuggestion = (options = {}) => {
  const inactivityMs = normalizeNonNegativeNumber(
    options.inactivityMs ?? 0,
    "inactivityMs",
  );
  const idleThresholdMs =
    options.idleThresholdMs === undefined || options.idleThresholdMs === null
      ? null
      : normalizeNonNegativeNumber(options.idleThresholdMs, "idleThresholdMs");

  return freezeRecord({
    source: normalizeNonEmptyToken(
      options.source ?? "runtime-inactivity-heuristic",
      "idle window suggestion source",
    ),
    suggestedPhase: normalizeNonEmptyToken(
      options.suggestedPhase ?? "idle",
      "idle window suggestion phase",
    ),
    inactivityMs,
    idleThresholdMs,
    thresholdReached:
      idleThresholdMs === null ? inactivityMs > 0 : inactivityMs >= idleThresholdMs,
    authorizesConsolidation: false,
    note: options.note ?? null,
  });
};

const createDefaultReserveWindowMs = (idleTriggerWindowMs) =>
  Math.min(
    Math.floor(idleTriggerWindowMs * 0.2),
    Math.max(idleTriggerWindowMs - 1, 0),
  );

export const createIdleWindowReconstructionBudget = (options = {}) => {
  if (!options || typeof options !== "object" || Array.isArray(options)) {
    throw new TypeError("idle window reconstruction budget options must be an object");
  }

  const idleTriggerWindowMs = normalizePositiveInteger(
    options.idleTriggerWindowMs,
    "idleTriggerWindowMs",
  );
  const requestedReserveWindowMs =
    options.reserveWindowMs ?? options.reconstructionReserveWindowMs;
  const reserveWindowMs = Math.min(
    requestedReserveWindowMs === undefined || requestedReserveWindowMs === null
      ? createDefaultReserveWindowMs(idleTriggerWindowMs)
      : normalizeNonNegativeInteger(
          requestedReserveWindowMs,
          "reserveWindowMs",
        ),
    Math.max(idleTriggerWindowMs - 1, 0),
  );

  return freezeRecord({
    source: "idle-trigger-window",
    idleTriggerWindowMs,
    reserveWindowMs,
    reconstructionBudgetMs: Math.max(1, idleTriggerWindowMs - reserveWindowMs),
  });
};

const normalizeIdleWindowSuggestion = (suggestion) => {
  if (!suggestion) {
    return null;
  }

  if (typeof suggestion !== "object") {
    throw new TypeError("idle window suggestion must be an object");
  }

  return createIdleWindowSuggestion(suggestion);
};

const getBlockedReason = (runtimePhase) => {
  if (!runtimePhase) {
    return "missing-runtime-phase";
  }

  if (!EXPLICIT_PHASE_AUTHORITIES.has(runtimePhase.authority)) {
    return "runtime-phase-not-caller-controlled";
  }

  if (!RUNTIME_AUTHORIZED_IDLE_PHASE_SET.has(runtimePhase.value)) {
    return "runtime-phase-not-idle-window";
  }

  return null;
};

export const evaluateIdleWindowAuthorization = (options = {}) => {
  const runtimePhase = normalizeRuntimePhase(options.runtimePhase);
  const blockedReason = getBlockedReason(runtimePhase);
  const eligible = blockedReason === null;
  const reconstructionBudget =
    options.idleTriggerWindowMs === undefined || options.idleTriggerWindowMs === null
      ? null
      : createIdleWindowReconstructionBudget({
          idleTriggerWindowMs: options.idleTriggerWindowMs,
          reserveWindowMs:
            options.reserveWindowMs ?? options.reconstructionReserveWindowMs,
        });

  return freezeRecord({
    agentId: normalizeAgentId(options.agentId),
    runtimePhase,
    inactivitySuggestion: normalizeIdleWindowSuggestion(
      options.inactivitySuggestion,
    ),
    reconstructionBudget,
    teamIdle: Boolean(options.teamIdle),
    eligible,
    opensConsolidation: eligible,
    decisionSource: eligible ? "runtime-phase" : null,
    blockedReason,
    requiresOfflineExecution: true,
  });
};

export const planIdleWindowConsolidation = (options = {}) => {
  const agents = options.agents ?? [];

  if (!Array.isArray(agents)) {
    throw new TypeError("agents must be an array");
  }

  const teamIdle = Boolean(options.teamIdle);
  const decisions = agents.map((agent) =>
    evaluateIdleWindowAuthorization({
      agentId: agent.agentId,
      runtimePhase: agent.runtimePhase,
      inactivitySuggestion: agent.inactivitySuggestion,
      idleTriggerWindowMs: agent.idleTriggerWindowMs,
      reserveWindowMs: agent.reserveWindowMs,
      reconstructionReserveWindowMs: agent.reconstructionReserveWindowMs,
      teamIdle,
    }),
  );

  const eligibleAgents = freezeRecordList(
    decisions.filter((decision) => decision.eligible),
  );
  const blockedAgents = freezeRecordList(
    decisions.filter((decision) => !decision.eligible),
  );

  return freezeRecord({
    teamIdle,
    windowAuthority: "runtime-phase",
    eligibleAgents,
    blockedAgents,
    eligibleCount: eligibleAgents.length,
    blockedCount: blockedAgents.length,
    batchWindowOpen: eligibleAgents.length > 0,
  });
};

File: go/runtime_phase.go
package brain

import "fmt"

var AuthorizedIdlePhases = []string{"idle", "rest", "break", "sleep"}

type RuntimePhase struct {
	Value     string `json:"value"`
	Authority string `json:"authority"`
	ChangedAt string `json:"changedAt,omitempty"`
	Note      string `json:"note,omitempty"`
}

type IdleWindowSuggestion struct {
	Source                  string `json:"source"`
	SuggestedPhase          string `json:"suggestedPhase"`
	InactivityMS            int64  `json:"inactivityMs"`
	IdleThresholdMS         int64  `json:"idleThresholdMs,omitempty"`
	ThresholdReached        bool   `json:"thresholdReached"`
	AuthorizesConsolidation bool   `json:"authorizesConsolidation"`
	Note                    string `json:"note,omitempty"`
}

type IdleWindowAuthorization struct {
	AgentID                  string                `json:"agentId"`
	RuntimePhase             *RuntimePhase         `json:"runtimePhase,omitempty"`
	InactivitySuggestion     *IdleWindowSuggestion `json:"inactivitySuggestion,omitempty"`
	TeamIdle                 bool                  `json:"teamIdle"`
	Eligible                 bool                  `json:"eligible"`
	OpensConsolidation       bool                  `json:"opensConsolidation"`
	DecisionSource           string                `json:"decisionSource,omitempty"`
	BlockedReason            string                `json:"blockedReason,omitempty"`
	RequiresOfflineExecution bool                  `json:"requiresOfflineExecution"`
}

type IdleWindowPlanAgent struct {
	AgentID              string
	RuntimePhase         any
	InactivitySuggestion *IdleWindowSuggestion
}

type IdleWindowConsolidationPlan struct {
	TeamIdle        bool                      `json:"teamIdle"`
	WindowAuthority string                    `json:"windowAuthority"`
	EligibleAgents  []IdleWindowAuthorization `json:"eligibleAgents"`
	BlockedAgents   []IdleWindowAuthorization `json:"blockedAgents"`
	EligibleCount   int                       `json:"eligibleCount"`
	BlockedCount    int                       `json:"blockedCount"`
	BatchWindowOpen bool                      `json:"batchWindowOpen"`
}

func CreateRuntimePhase(value string, authority string, changedAt string, note string) (RuntimePhase, error) {
	normalizedValue, err := normalizeToken(value, "runtime phase")
	if err != nil {
		return RuntimePhase{}, err
	}
	if authority == "" {
		authority = "caller"
	}
	normalizedAuthority, err := normalizeToken(authority, "runtime phase authority")
	if err != nil {
		return RuntimePhase{}, err
	}
	normalizedChangedAt, err := normalizeTimeString(changedAt)
	if err != nil {
		return RuntimePhase{}, fmt.Errorf("changedAt must be RFC3339: %w", err)
	}
	return RuntimePhase{
		Value:     normalizedValue,
		Authority: normalizedAuthority,
		ChangedAt: normalizedChangedAt,
		Note:      normalizeOptionalString(note),
	}, nil
}

func CreateIdleWindowSuggestion(source string, suggestedPhase string, inactivityMS int64, idleThresholdMS int64, note string) (IdleWindowSuggestion, error) {
	if source == "" {
		source = "runtime-inactivity-heuristic"
	}
	normalizedSource, err := normalizeToken(source, "idle window suggestion source")
	if err != nil {
		return IdleWindowSuggestion{}, err
	}
	if suggestedPhase == "" {
		suggestedPhase = "idle"
	}
	normalizedPhase, err := normalizeToken(suggestedPhase, "idle window suggestion phase")
	if err != nil {
		return IdleWindowSuggestion{}, err
	}
	if inactivityMS < 0 || idleThresholdMS < 0 {
		return IdleWindowSuggestion{}, fmt.Errorf("idle window timing must be non-negative")
	}
	return IdleWindowSuggestion{
		Source:                  normalizedSource,
		SuggestedPhase:          normalizedPhase,
		InactivityMS:            inactivityMS,
		IdleThresholdMS:         idleThresholdMS,
		ThresholdReached:        idleThresholdMS == 0 || inactivityMS >= idleThresholdMS,
		AuthorizesConsolidation: false,
		Note:                    normalizeOptionalString(note),
	}, nil
}

func normalizeRuntimePhaseInput(input any) (*RuntimePhase, error) {
	switch value := input.(type) {
	case nil:
		return nil, nil
	case RuntimePhase:
		copied := value
		return &copied, nil
	case *RuntimePhase:
		if value == nil {
			return nil, nil
		}
		copied := *value
		return &copied, nil
	case string:
		created, err := CreateRuntimePhase(value, "caller", "", "")
		if err != nil {
			return nil, err
		}
		return &created, nil
	default:
		return nil, fmt.Errorf("unsupported runtime phase input %T", input)
	}
}

func evaluateBlockedReason(runtimePhase *RuntimePhase) string {
	if runtimePhase == nil {
		return "missing-runtime-phase"
	}
	if runtimePhase.Authority != "caller" {
		return "runtime-phase-not-caller-controlled"
	}
	if !containsString(AuthorizedIdlePhases, runtimePhase.Value) {
		return "runtime-phase-not-idle-window"
	}
	return ""
}

func EvaluateIdleWindowAuthorization(agentID string, runtimePhase any, inactivitySuggestion *IdleWindowSuggestion, teamIdle bool) (IdleWindowAuthorization, error) {
	normalizedAgentID, err := normalizeRequiredString(agentID, "agentId")
	if err != nil {
		return IdleWindowAuthorization{}, err
	}
	normalizedPhase, err := normalizeRuntimePhaseInput(runtimePhase)
	if err != nil {
		return IdleWindowAuthorization{}, err
	}
	blockedReason := evaluateBlockedReason(normalizedPhase)
	eligible := blockedReason == ""
	return IdleWindowAuthorization{
		AgentID:                  normalizedAgentID,
		RuntimePhase:             normalizedPhase,
		InactivitySuggestion:     inactivitySuggestion,
		TeamIdle:                 teamIdle,
		Eligible:                 eligible,
		OpensConsolidation:       eligible,
		DecisionSource:           map[bool]string{true: "runtime-phase", false: ""}[eligible],
		BlockedReason:            blockedReason,
		RequiresOfflineExecution: true,
	}, nil
}

func PlanIdleWindowConsolidation(teamIdle bool, agents []IdleWindowPlanAgent) (IdleWindowConsolidationPlan, error) {
	plan := IdleWindowConsolidationPlan{
		TeamIdle:        teamIdle,
		WindowAuthority: "runtime-phase",
		EligibleAgents:  []IdleWindowAuthorization{},
		BlockedAgents:   []IdleWindowAuthorization{},
	}
	for _, agent := range agents {
		decision, err := EvaluateIdleWindowAuthorization(agent.AgentID, agent.RuntimePhase, agent.InactivitySuggestion, teamIdle)
		if err != nil {
			return IdleWindowConsolidationPlan{}, err
		}
		if decision.Eligible {
			plan.EligibleAgents = append(plan.EligibleAgents, decision)
		} else {
			plan.BlockedAgents = append(plan.BlockedAgents, decision)
		}
	}
	plan.EligibleCount = len(plan.EligibleAgents)
	plan.BlockedCount = len(plan.BlockedAgents)
	plan.BatchWindowOpen = plan.EligibleCount > 0
	return plan, nil
}
